src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -954,132 +955,113 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     // Appends end-user visible description of `place` to `buf`.
     fn append_place_to_string(
         &self,
-        place: &Place<'tcx>,
+        place: &NeoPlace<'tcx>,
         buf: &mut String,
-        mut autoderef: bool,
         including_downcast: &IncludingDowncast,
     ) -> Result<(), ()> {
-        match *place {
-            Place::Promoted(_) => {
-                buf.push_str("promoted");
+        self.append_place_base_to_string(&place.base, buf)?;
```

Comment:
This logic isn't really right, as you can see from the test failures:

```diff
- error[E0501]: cannot borrow `*a` as mutable because previous closure requires unique access
+ error[E0501]: cannot borrow `a*` as mutable because previous closure requires unique access
```

My inclination would be code this up as a "tree-walking" case that works from the *end* of the projections backwards. In particular you need to be careful to handle precedence and parentheses correctly.

In particular, if you find yourself with a field projection where the prior projection is a deref, then you want to wrap the deref in parentheses. So `(*x).f` — this is because `*x.f` means `*(x.f)`.